src/store.c

```diff
@@ -339,144 +374,255 @@ static void writer_close(
     close(store->fd);
 }
 
+static void init_store_offsets(struct rill_store* store, uint64_t pairs)
+{
+    store->head->index_b_off = store->head->index_a_off + indexer_cap(pairs);
+    store->head->data_a_off = store->head->index_b_off + indexer_cap(pairs);
+
+    store->index_a = (void *) ((uintptr_t) store->vma + store->head->index_a_off);
+    store->index_b = (void *) ((uintptr_t) store->vma + store->head->index_b_off);
+    store->data_a = (void *) ((uintptr_t) store->vma + store->head->data_a_off);
+}
+
+static void prepare_col_b_offsets(
+    struct rill_store* store, struct encoder* coder_a)
+{
+    store->head->data_b_off = store->head->data_a_off + coder_off(coder_a);
+    store->data_b = (void *) ((uintptr_t) store->vma + store->head->data_b_off);
+}
+
 bool rill_store_write(
         const char *file,
-        rill_ts_t ts, size_t quant,
+        rill_ts_t ts,
+        size_t quant,
         struct rill_pairs *pairs)
 {
     rill_pairs_compact(pairs);
     if (!pairs->len) return true;
 
     struct vals *vals = vals_from_pairs(pairs);
     if (!vals) goto fail_vals;
+    struct vals *invert_vals = keys_from_pairs(pairs); // TODO fix fail cleanup
 
     struct rill_store store = {0};
-    if (!writer_open(&store, file, vals, pairs->len, ts, quant)) {
+    if (!writer_open(&store, file, vals, invert_vals,
+                     pairs->len, ts, quant)) {
         rill_fail("unable to create '%s'", file);
         goto fail_open;
     }
 
-    struct indexer *indexer = indexer_alloc(pairs->len);
-    struct encoder coder = writer_begin(&store, vals, indexer);
+    struct indexer *indexer_a = indexer_alloc(pairs->len); // TODO catch fail alloc here
+    if (!indexer_a) goto fail_indexer_a_alloc;
+    struct indexer *indexer_b = indexer_alloc(pairs->len); // TODO catch fail alloc here
+    if (!indexer_b) goto fail_indexer_b_alloc;
```

Comment:
Note how the indexer struct is basically the same as the index struct used in the mmap. Since you're already pre-allocated space in the mmap region, you should be able to just use that directly instead of allocating another index which you then have to write out.